wave_containers/include/wave/containers/impl/measurement_container.hpp

```diff
@@ -138,48 +129,35 @@ typename MeasurementContainer<T>::iterator MeasurementContainer<T>::erase(
 template <typename T>
 typename MeasurementContainer<T>::ValueType MeasurementContainer<T>::get(
   const TimeType &t, const SensorIdType &s) const {
-    // To interpolate measurements from one sensor only, we first extract all
-    // measurements from that sensor, then search by time. This method is based
-    // on one described here:
-    // http://www.boost.org/doc/libs/1_63_0/libs/multi_index/doc/examples.html#example6
-
-    // Find all measurements from this sensor
+    // To interpolate measurements from one sensor only, we use the index sorted
+    // by sensor_id first
     const auto &sensor_index = this->storage.template get<
       typename internal::measurement_container<T>::sensor_index>();
-    const auto sensor_it = sensor_index.equal_range(s);
-
-    // Construct a view, indexed by time, with only those measurements
-    auto time_view = typename internal::measurement_container<T>::time_view{};
-    for (auto it = sensor_it.first; it != sensor_it.second; ++it) {
-        time_view.insert(&*it);  // the view holds pointers to the measurements
-    }
 
-    // find the first element with time >= t
-    auto i_next = time_view.lower_bound(t);
+    // find the first element with sensor>=s and time >= t
+    auto i_next = sensor_index.lower_bound(boost::make_tuple(s, t));
 
-    if (i_next == time_view.end()) {
+    if (i_next == sensor_index.end() || i_next->sensor_id != s) {
         // Requested time is not between two measurements for this sensor
         throw std::out_of_range("MeasurementContainer::get");
     }
 
-    const auto p_next = *i_next;
-
-    if (t == p_next->time_point) {
+    if (t == i_next->time_point) {
         // Requested time exactly matches
-        return p_next->value;
+        return i_next->value;
     }
 
-    // If no exact match, need at least one more measurement to interpolate
-    if (i_next == time_view.begin()) {
+    // If no exact match, need at least one previous measurement to interpolate
+    const auto i_prev = std::prev(i_next);
+
+    if (i_next == sensor_index.begin() || i_prev->sensor_id != s) {
```

Comment:
Also does this still hold? Because now the sensor could very well be the only sensor in the `sensor_index`?

Reply:
Maybe I don't understand what the sensor view is. It's the list of all measurements for the sensor `s`, correct? Would there be a chance that `i_prev` has a sensor id that isn't `s`?

Reply:
There is no sensor view now. sensor_index has all the same measurements, just sorted by sensor first. Yes there is a chance.